pkg/controllers/provisioning/provisioner.go

```diff
@@ -17,128 +17,148 @@ package provisioning
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/util/workqueue"
 	"knative.dev/pkg/logging"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	crmetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
 )
 
 var (
-	MaxBatchDuration = time.Second * 10
-	MinBatchDuration = time.Second * 1
-	// MaxPodsPerBatch limits the number of pods we process at one time to avoid using too much memory
-	MaxPodsPerBatch = 2_000
+	defaultBatchOptions = BatchOptions{
+		IdleDuration: 1 * time.Second,
+		MaxDuration:  10 * time.Second,
+		MaxSize:      2_000,
+	}
 )
 
 // Provisioner waits for enqueued pods, batches them, creates capacity and binds the pods to the capacity.
 type Provisioner struct {
 	// State
 	*v1alpha5.Provisioner
-	pods    chan *v1.Pod
-	results chan error
-	done    <-chan struct{}
+	Batches sync.Map
+	Ctx     context.Context
 	Stop    context.CancelFunc
 	// Dependencies
 	cloudProvider cloudprovider.CloudProvider
 	kubeClient    client.Client
-	scheduler     *scheduling.Scheduler
-	launcher      *Launcher
-}
-
-func (p *Provisioner) Start(ctx context.Context) {
-	go func() {
-		logging.FromContext(ctx).Info("Starting provisioner")
-		for {
-			select {
-			case <-p.done:
-				logging.FromContext(ctx).Info("Stopping provisioner")
-				return
-			default:
-				if err := p.provision(ctx); err != nil {
-					logging.FromContext(ctx).Errorf("Provisioning failed, %s", err.Error())
-				}
-			}
-		}
-	}()
+	coreV1Client  corev1.CoreV1Interface
+	packer        *binpacking.Packer
 }
 
-func (p *Provisioner) provision(ctx context.Context) (err error) {
-	// Wait for a batch of pods
-	pods := p.Batch(ctx)
-	// Communicate the result of the provisioning loop to each of the pods.
-	defer func() {
-		for i := 0; i < len(pods); i++ {
-			select {
-			case p.results <- err: // Block until result is communicated
-			case <-p.done: // Leave if closed
-			}
-		}
-	}()
-	// Separate pods by scheduling constraints
-	schedules, err := p.scheduler.Solve(ctx, p.Provisioner, pods)
-	if err != nil {
-		return fmt.Errorf("solving scheduling constraints, %w", err)
-	}
-	// Get Instance Types, offering availability may vary over time
-	instanceTypes, err := p.cloudProvider.GetInstanceTypes(ctx, &p.Spec.Constraints)
+// Add (threadsafe) the pod to the provisioner and block until it's processed.
+// The caller is responsible for verifying that the pod was scheduled correctly.
+func (p *Provisioner) Add(ctx context.Context, pod *v1.Pod) error {
+	// Fail if constraints are not viable for pod
+	constraints, err := p.Spec.Constraints.Tighten(pod)
 	if err != nil {
-		return fmt.Errorf("getting instance types")
+		return err
 	}
-	// Launch capacity and bind pods
-	if err := p.launcher.Launch(ctx, p.Provisioner, schedules, instanceTypes); err != nil {
-		return fmt.Errorf("launching capacity, %w", err)
+
+	// Atomically create or use an existing batch.
+	value, ok := p.Batches.LoadOrStore(constraints.MustHash(), NewBatch(ctx, defaultBatchOptions))
```

Comment:
Why do we create multiple batches here? When we batch pods according to constraints hash, doesn't it limited the number of pods that can be batched together? It seems to me this is a bit redundant given that the packer.Pack also groups pods according to constraints hash.

Reply:
I am also a little bit worry that this constraints-hash grouping mechanism will further limit ourselves for later if we want to try grouping pods with compatible constraints.

Reply:
>It seems to me this is a bit redundant given that the packer.Pack also groups pods according to constraints hash.

Binpacking and scheduling are doing different things. In binpacking it (currently) computes a set of alternatives (OR). In scheduling, it's tightening a set of rules (AND).

> I am also a little bit worry that this constraints-hash grouping mechanism will further limit ourselves for later if we want to try grouping pods with compatible constraints.

This is a very good point -- as long as this mechanism is async we're going to have race conditions that will require retry. The atomicity here likely isn't helping as much as I think it is.

Reply:
> Why do we create multiple batches here?
We need multiple batches because the constraints may not be the same for (podB + provisionerA) and (podB + provisionerA).

Reply:
> 
So now the multiple batches is responsible of grouping pods according to their constraints. I like that. Much cleaner.